src/java.base/share/classes/javax/crypto/KDF.java

```diff
@@ -0,0 +1,669 @@
+ * <i>delayed provider selection</i>. The primary reason this is done is to
+ * ensure that the selected provider can handle the key material that is passed
+ * to those methods - for example, the key material may reside on a hardware
+ * device that only a specific {@code KDF} provider can utilize. The {@code
+ * getInstance} method returns a {@code KDF} object as long as there exists
+ * at least one registered security provider that implements the algorithm
+ * and supports the optional parameters. The delayed provider selection
+ * process traverses the list of registered security providers, starting with
+ * the most preferred {@code Provider}. The first provider that supports the
+ * specified algorithm, optional parameters, and key material is selected.
+ * <p>
+ * If the {@code getProviderName} or {@code getParameters} method is called
+ * before the {@code deriveKey} or {@code deriveData} methods, the first
+ * provider supporting the {@code KDF} algorithm and optional
+ * {@code KDFParameters} is chosen. This provider may not support the key
+ * material that is subsequently passed to the {@code deriveKey} or
+ * {@code deriveData} methods. Therefore, it is recommended not to call the
+ * {@code getProviderName} or {@code getParameters} methods until after a key
+ * derivation operation. Once a provider is selected, it cannot be changed.
+ *
+ * @see KDFParameters
+ * @see SecretKey
+ * @since 24
+ */
+@PreviewFeature(feature = PreviewFeature.Feature.KEY_DERIVATION)
+public final class KDF {
+
+    private static final Debug pdebug = Debug.getInstance("provider",
+                                                          "Provider");
+    private static final boolean skipDebug = Debug.isOn("engine=")
+                                             && !Debug.isOn("kdf");
+
+    private record Delegate(KDFSpi spi, Provider provider) {}
+
+    //guarded by 'lock'
+    private Delegate theOne;
+    //guarded by 'lock'
+    private Delegate candidate;
+
+    // The name of the KDF algorithm.
+    private final String algorithm;
+
+    // Additional KDF configuration parameters
+    private final KDFParameters kdfParameters;
+
+    // remaining services to try in provider selection
+    // null once provider is selected
+    private final Iterator<Service> serviceIterator;
+
+    // This lock provides mutual exclusion, preventing multiple threads from
+    // concurrently initializing the same instance (delayed provider selection)
+    // in a way which would corrupt the internal state.
+    private final Object lock = new Object();
+
+
+    // Instantiates a {@code KDF} object. This constructor is called when a
+    // provider is supplied to {@code getInstance}.
+    //
+    // @param delegate the delegate
+    // @param algorithm the algorithm
+    // @param kdfParameters the parameters
+    private KDF(Delegate delegate, String algorithm,
+                KDFParameters kdfParameters) {
```

Comment:
There is really no need to pass in `kdfParameters` in this case. Set `this.kdfParameters` to `null`.